presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/client/AwsRequestSigner.java

```diff
@@ -0,0 +1,115 @@
+import java.io.InputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.lang.String.CASE_INSENSITIVE_ORDER;
+import static org.apache.http.protocol.HttpCoreContext.HTTP_TARGET_HOST;
+
+class AwsRequestSigner
+        implements HttpRequestInterceptor
+{
+    private static final String SERVICE_NAME = "es";
+    private final AWSCredentialsProvider credentialsProvider;
+    private final AWS4Signer signer;
+
+    public AwsRequestSigner(String region, AWSCredentialsProvider credentialsProvider)
+    {
+        this.credentialsProvider = credentialsProvider;
+        this.signer = new AWS4Signer();
+
+        signer.setServiceName(SERVICE_NAME);
+        signer.setRegionName(region);
+    }
+
+    @Override
+    public void process(final HttpRequest request, final HttpContext context)
+            throws IOException
+    {
+        String method = request.getRequestLine().getMethod();
+
+        URI uri = URI.create(request.getRequestLine().getUri());
+        URIBuilder uriBuilder = new URIBuilder(uri);
+
+        Map<String, List<String>> parameters = new TreeMap<>(CASE_INSENSITIVE_ORDER);
+        for (NameValuePair parameter : uriBuilder.getQueryParams()) {
+            parameters.computeIfAbsent(parameter.getName(), key -> new ArrayList<>())
+                    .add(parameter.getValue());
+        }
+
+        Map<String, String> headers = Arrays.stream(request.getAllHeaders())
+                .collect(toImmutableMap(Header::getName, Header::getValue));
+
+        InputStream content = null;
+        if (request instanceof HttpEntityEnclosingRequest) {
+            HttpEntityEnclosingRequest enclosingRequest = (HttpEntityEnclosingRequest) request;
+            if (enclosingRequest.getEntity() != null) {
+                content = enclosingRequest.getEntity().getContent();
+            }
+        }
+
+        DefaultRequest<?> awsRequest = new DefaultRequest<>(SERVICE_NAME);
+
+        HttpHost host = (HttpHost) context.getAttribute(HTTP_TARGET_HOST);
+        if (host != null) {
+            awsRequest.setEndpoint(URI.create(host.toURI()));
+        }
+        awsRequest.setHttpMethod(HttpMethodName.fromValue(method));
+        awsRequest.setResourcePath(uri.getRawPath());
+        awsRequest.setContent(content);
+        awsRequest.setParameters(parameters);
+        awsRequest.setHeaders(headers);
+
+        signer.sign(awsRequest, credentialsProvider.getCredentials());
+
+        Header[] newHeaders = awsRequest.getHeaders().entrySet().stream()
+                .map(entry -> new BasicHeader(entry.getKey(), entry.getValue()))
+                .toArray(Header[]::new);
+
+        request.setHeaders(newHeaders);
+
+        InputStream newContent = awsRequest.getContent();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *newContent*. Currently, there are execution paths that do not contain closure statements, for example, when AwsRequestSigner.checkState throws an exception. Ensure *newContent* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).